src/thermo/BinarySolutionTabulatedThermo.cpp

```diff
@@ -211,27 +258,72 @@ void BinarySolutionTabulatedThermo::initThermoXML(XML_Node& phaseNode, const std
     ThermoPhase::initThermoXML(phaseNode, id_);
 }
 
-std::pair<double, double> BinarySolutionTabulatedThermo::interpolate(double x) const
+double BinarySolutionTabulatedThermo::interpolate(double x,const vector_fp& molefrac,const vector_fp& inputData) const
 {
-    std::pair<double, double> c;
+    double c;
     // Check if x is out of bound
-    if (x > m_molefrac_tab.back()) {
-        c.first = m_enthalpy_tab.back();
-        c.second = m_entropy_tab.back();
+    if (x > molefrac.back()) {
+        c= inputData.back();
         return c;
     }
-    if (x < m_molefrac_tab[0]) {
-        c.first = m_enthalpy_tab[0];
-        c.second = m_entropy_tab[0];
+    if (x < molefrac.front()) {
+        c= inputData.front();
         return c;
     }
-    size_t i = std::distance(m_molefrac_tab.begin(),
-      std::lower_bound(m_molefrac_tab.begin(), m_molefrac_tab.end(), x));
-    c.first = m_enthalpy_tab[i-1] + (m_enthalpy_tab[i] - m_enthalpy_tab[i-1])
-      * (x - m_molefrac_tab[i-1])/(m_molefrac_tab[i]- m_molefrac_tab[i-1]);
-    c.second = m_entropy_tab[i-1] + (m_entropy_tab[i] - m_entropy_tab[i-1])
-      * (x - m_molefrac_tab[i-1])/(m_molefrac_tab[i]- m_molefrac_tab[i-1]);
+    size_t i = std::distance(molefrac.begin(),
+      std::lower_bound(molefrac.begin(), molefrac.end(), x));
+    c = inputData[i-1] + (inputData[i] - inputData[i-1])
+      * (x - molefrac[i-1])/(molefrac[i]- molefrac[i-1]);
     return c;
 }
 
+void BinarySolutionTabulatedThermo::integrate(vector_fp& inputData,vector_fp& integratedData,vector_fp& moleFraction)
+{
+	integratedData.resize(inputData.size());
+	for(size_t i = 0; i < integratedData.size(); i++){
+		if (i==0 and moleFraction[0]==0){
+			integratedData[i]=0;
+		}
+		else if (i==0 and moleFraction[0]>0){
+			integratedData[i]=moleFraction[i]*inputData[0];
+		}
```

Comment:
I think you can avoid this conditional by doing the first element outside the loop and starting the loop at 1 instead of 0. Also, these two conditions are identical, since if `moleFraction[i] == 0` is true, then `moleFraction[i]*inputData[i]` will also be 0. There's no need to have the condition, I think you can just do the multiplication.